torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1263,6 +1264,59 @@ def no_sync(self) -> Generator:
                     "`True` while in the `no_sync()` context manager"
                 m._require_backward_grad_sync = old_flag
 
+    @torch.no_grad()
+    def _clip_grad_norm_(
+        self, max_norm: Union[float, int], norm_type: Union[float, int] = 2.0
+    ) -> None:
+        """
+        Clip all gradients at this point in time. The norm is computed over all
+        gradients together, as if they were concatenated into a single vector.
+        Gradients are modified in-place.
+
+        Args:
+            max_norm (float or int): max norm of the gradients
+            norm_type (float or int): type of the used p-norm. Can be ``'inf'``
+                for infinity norm.
+
+        Returns:
+            Total norm of the parameters (viewed as a single vector).
+
+        .. note:: This is analogous to `torch.nn.utils.clip_grad_norm_` but
+            handles the partitioning and multiple devices per rank under the
+            hood. The default torch util is not applicable here, because each
+            rank only has a partial view of all the grads in the model, so
+            calling it in the OSS context would lead to different scaling being
+            applied per subset of model parameters.
+
+        .. warning:: This needs to be called on all ranks, since synchronization
+            primitives will be used.
+        """
+        self._lazy_init()
+        assert self._is_root, "clip_grad_norm should only be called on the root (parent) instance"
+        self._assert_state(TrainingState_.IDLE)
+
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+        # Computes the max norm for this shard's gradients and sync's across workers
+        local_norm = _calc_grad_norm(self.params, norm_type).cuda()
```

Comment:
why do we need to call `.cuda()` here? Seems like `_calc_grad_norm` should return the value on the same device as parameters in self.params?

Reply:
Right, but when the cpu_offload is set true, the params are on CPU. So looks like we do need this here otherwise, the test fails when calling the c10d API.

Reply:
I see, makes sense. This implicitly assumes user has specified this rank's device appropriately with `torch.cuda.set_device()` which they should do, but could you add a small comment here anyways to give potential debugging hints to future devs :)

Reply:
I think it is because backend of process group is NCCL, and all_reduce below expects CUDA tensors. yes, add a comment here will be great

Reply:
sure.